From 4a6277a70e3239c966815a3f92ebea3c6baa41bf Mon Sep 17 00:00:00 2001
Message-Id: <4a6277a70e3239c966815a3f92ebea3c6baa41bf.1334581530.git.minovotn@redhat.com>
In-Reply-To: <28774cb15d2443e325049ef410956f5635a1484b.1334581530.git.minovotn@redhat.com>
References: <28774cb15d2443e325049ef410956f5635a1484b.1334581530.git.minovotn@redhat.com>
From: Gerd Hoffmann <kraxel@redhat.com>
Date: Fri, 30 Mar 2012 12:35:34 +0200
Subject: [PATCH 4/6] usb: use USBDescriptor for interface descriptors.

RH-Author: Gerd Hoffmann <kraxel@redhat.com>
Message-id: <1333110936-22700-5-git-send-email-kraxel@redhat.com>
Patchwork-id: 39044
O-Subject: [RHEL-6.3 qemu-kvm PATCH 4/6] usb: use USBDescriptor for interface descriptors.
Bugzilla: 807878
RH-Acked-by: Hans de Goede <hdegoede@redhat.com>
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

Add interface descriptor substruct to USBDescriptor,
use it in the descriptor generator code.

upstream: http://patchwork.ozlabs.org/patch/149416/

Signed-off-by: Gerd Hoffmann <kraxel@redhat.com>
---
 hw/usb-desc.c |   20 +++++++++++---------
 hw/usb-desc.h |    9 +++++++++
 2 files changed, 20 insertions(+), 9 deletions(-)

Signed-off-by: Michal Novotny <minovotn@redhat.com>
---
 hw/usb-desc.c |   20 +++++++++++---------
 hw/usb-desc.h |    9 +++++++++
 2 files changed, 20 insertions(+), 9 deletions(-)

diff --git a/hw/usb-desc.c b/hw/usb-desc.c
index deb1903..6a13dfa 100644
--- a/hw/usb-desc.c
+++ b/hw/usb-desc.c
@@ -159,20 +159,22 @@ int usb_desc_iface(const USBDescIface *iface, uint8_t *dest, size_t len)
 {
     uint8_t bLength = 0x09;
     int i, rc, pos = 0;
+    USBDescriptor *d = (void *)dest;
 
     if (len < bLength) {
         return -1;
     }
 
-    dest[0x00] = bLength;
-    dest[0x01] = USB_DT_INTERFACE;
-    dest[0x02] = iface->bInterfaceNumber;
-    dest[0x03] = iface->bAlternateSetting;
-    dest[0x04] = iface->bNumEndpoints;
-    dest[0x05] = iface->bInterfaceClass;
-    dest[0x06] = iface->bInterfaceSubClass;
-    dest[0x07] = iface->bInterfaceProtocol;
-    dest[0x08] = iface->iInterface;
+    d->bLength                        = bLength;
+    d->bDescriptorType                = USB_DT_INTERFACE;
+
+    d->u.interface.bInterfaceNumber   = iface->bInterfaceNumber;
+    d->u.interface.bAlternateSetting  = iface->bAlternateSetting;
+    d->u.interface.bNumEndpoints      = iface->bNumEndpoints;
+    d->u.interface.bInterfaceClass    = iface->bInterfaceClass;
+    d->u.interface.bInterfaceSubClass = iface->bInterfaceSubClass;
+    d->u.interface.bInterfaceProtocol = iface->bInterfaceProtocol;
+    d->u.interface.iInterface         = iface->iInterface;
     pos += bLength;
 
     for (i = 0; i < iface->ndesc; i++) {
diff --git a/hw/usb-desc.h b/hw/usb-desc.h
index 2f09101..2364605 100644
--- a/hw/usb-desc.h
+++ b/hw/usb-desc.h
@@ -45,6 +45,15 @@ typedef struct USBDescriptor {
             uint8_t           bmAttributes;
             uint8_t           bMaxPower;
         } config;
+        struct {
+            uint8_t           bInterfaceNumber;
+            uint8_t           bAlternateSetting;
+            uint8_t           bNumEndpoints;
+            uint8_t           bInterfaceClass;
+            uint8_t           bInterfaceSubClass;
+            uint8_t           bInterfaceProtocol;
+            uint8_t           iInterface;
+        } interface;
     } u;
 } __attribute__((packed)) USBDescriptor;
 
-- 
1.7.7.6

